Source/JavaScriptCore/runtime/WaiterListManager.cpp

```diff
@@ -0,0 +1,289 @@
+
+        return;
+    }
+
+    // If waiter is a SyncWaiter, we null out it's vn to indicate that this waiter
+    // is removed from the WaiterList.
+    ASSERT(waiter->vm());
+    waiter->nullOutVM();
+    waiter->getCondition().notifyOne();
+}
+
+size_t WaiterListManager::waiterListSize(void* ptr)
+{
+    RefPtr<WaiterList> list = findList(ptr);
+    size_t size = 0;
+    if (list) {
+        Locker listLocker { list->lock };
+        size = list->size();
+    }
+    return size;
+}
+
+void WaiterListManager::cancelAsyncWaiter(Waiter* waiter)
+{
+    ASSERT(waiter->isAsync());
+    VM& vm = *waiter->vm();
+    bool didCancelAndRemove = vm.deferredWorkTimer->cancelAndRemovePendingWork(waiter->getTicket());
+    ASSERT_UNUSED(didCancelAndRemove, didCancelAndRemove);
+    waiter->nullOutTicket();
+    waiter->cancelTimer();
+}
+
+void WaiterListManager::unregisterVM(VM* vm)
+{
+    Locker waiterListsLocker { m_waiterListsLock };
+    for (auto& entry : m_waiterLists) {
+        WaiterList* list = entry.value.get();
+        if (list) {
+            Locker listLocker { list->lock };
+            list->removeIf(listLocker, [&](Waiter* waiter) {
+                if (waiter->vm() == vm) {
+                    // This is weird. If the vm is about destructing, then
+                    // it shouldn't been blocked or have any deferred work left.
+                    // That means we shouldn't find any SyncWaiter or AsyncWaiter
+                    // associating to the vm. Maybe we should throw error or add assertion here?
+                    if (waiter->isAsync())
+                        cancelAsyncWaiter(waiter);
+                    else
+                        UNREACHABLE();
+
+                    dataLogLnIf(WaiterListsManagerInternal::verbose,
+                        "WaiterListManager::unregisterVM ",
+                        (waiter->isAsync() ? " deleted AsyncWaiter " : " removed SyncWaiter "),
+                        RawPointer(waiter),
+                        " in WaiterList for ptr ",
+                        RawPointer(entry.key));
+
+                    return true;
+                }
+                return false;
+            });
+        }
+    }
+}
+
+void WaiterListManager::unregisterSharedArrayBuffer(uint8_t* arrayPtr, size_t size)
+{
+    Locker listLocker { m_waiterListsLock };
+    m_waiterLists.removeIf([&](auto& entry) {
+        if (entry.key >= arrayPtr && entry.key < arrayPtr + size) {
+            WaiterList* list = entry.value.get();
+            if (list) {
+                Locker listLocker { list->lock };
+                list->removeIf(listLocker, [&](Waiter* waiter) {
+                    if (waiter->isAsync()) {
+                        // If the AsyncWaiter has valid timer, then let it
+                        // timeout. Otherwise un-task it.
+                        if (waiter->hasTimer())
+                            return false;
+                        cancelAsyncWaiter(waiter);
```

Comment:
Discussed offline. We should not allow keeping waiters in the list here. We must remove waiters always.